// binarykvparser/index.js
var ByteBuffer = require('bytebuffer');

var Type = {
	None: 0,
	String: 1,
	Int32: 2,
	Float32: 3,
	Pointer: 4,
	WideString: 5,
	Color: 6,
	UInt64: 7,
	End: 8
};

exports.parse = function(buffer) {
	if(!ByteBuffer.isByteBuffer(buffer)) {
		buffer = ByteBuffer.wrap(buffer, 'utf8', true);
	}
	
	var obj = {};
	var type, name, value;
	
	while(true) {
		type = buffer.readByte();
		
		if(type == Type.End) {
			break;
		}
		
		name = buffer.readCString();
		
		if(type === Type.None && !name && !Object.keys(obj).length) {
			// Root node
			name = buffer.readCString();
		}
		
		switch(type) {
			case Type.None:
				value = exports.parse(buffer);
				break;
			
			case Type.String:
				value = buffer.readCString();
				break;
			
			case Type.Int32:
			case Type.Color:
			case Type.Pointer:
				value = buffer.readInt32();
				break;
			
			case Type.UInt64:
				value = buffer.readUint64();
				break;
			
			case Type.Float32:
				value = buffer.readFloat32();
				break;
			
			default:
				throw new Error("Unknown KV type " + type + " encountered at offset " + buffer.offset);
		}
		
		if(name) {
			obj[name] = convertObject(value);
		}
	}
	
	return obj;
}

/**
 * Converts an object to an array if it's an array-like object
 * @param Object obj
 * @returns Object|Array
 */
function convertObject(obj) {
	if(typeof obj !== 'object') {
		return obj;
	}

	var keys = Object.keys(obj);

	var i;
	for(i = 0; i < keys.length; i++) {
		keys[i] = parseInt(keys[i], 10);
		if(isNaN(keys[i])) {
			return obj;
		}
	}

	keys.sort(function(a, b) {
		if(a == b) {
			return 0;
		} else {
			return a < b ? -1 : 1;
		}
	});

	for(i = 0; i < keys.length; i++) {
		if(keys[i] != i) {
			return obj;
		}
	}

	obj.length = keys.length;
	return Array.prototype.slice.call(obj);
}
